bindings/src/main/java/io/pravega/storage/filesystem/FileSystemStorage.java

```diff
@@ -303,28 +303,29 @@ private Void doWrite(SegmentHandle handle, long offset, InputStream data, int le
             throw new StreamSegmentSealedException(handle.getSegmentName());
         }
 
-        long fileSize = path.toFile().length();
-        if (fileSize < offset) {
-            throw new BadOffsetException(handle.getSegmentName(), fileSize, offset);
-        } else {
-            long totalBytesWritten = 0;
-            try (FileChannel channel = FileChannel.open(path, StandardOpenOption.WRITE)) {
-                // Wrap the input data into a ReadableByteChannel, but do not close it. Doing so will result in closing
-                // the underlying InputStream, which is not desirable if it is to be reused.
-                ReadableByteChannel sourceChannel = Channels.newChannel(data);
-                while (length != 0) {
-                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
-                    assert bytesWritten > 0 : "Unable to make any progress transferring data.";
-                    offset += bytesWritten;
-                    totalBytesWritten += bytesWritten;
-                    length -= bytesWritten;
-                }
+        long totalBytesWritten = 0;
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.WRITE)) {
+            long fileSize = channel.size();
+            if (fileSize != offset) {
+                throw new BadOffsetException(handle.getSegmentName(), fileSize, offset);
             }
-            FileSystemMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
-            FileSystemMetrics.WRITE_BYTES.add(totalBytesWritten);
-            LoggerHelpers.traceLeave(log, "write", traceId);
-            return null;
+
+            // Wrap the input data into a ReadableByteChannel, but do not close it. Doing so will result in closing
+            // the underlying InputStream, which is not desirable if it is to be reused.
+            ReadableByteChannel sourceChannel = Channels.newChannel(data);
+            while (length != 0) {
+                long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                assert bytesWritten > 0 : "Unable to make any progress transferring data.";
+                offset += bytesWritten;
+                totalBytesWritten += bytesWritten;
+                length -= bytesWritten;
+            }
+            channel.force(false);
```

Comment:
Ok, we are going with only forcing the metadata update in the case it is strictly necessary, which we are delegating to the underlying OS implementation, yes?

Reply:
Yes. In this case the underlying OS will most sync metadata. (Which will happen most of the time with mostly append-only writes in our case.)